internal/cluster/distribution/pke/pkeaws/pkeawsworkflow/activity_nlb_health_check.go

```diff
@@ -0,0 +1,143 @@
+// Copyright © 2020 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pkeawsworkflow
+
+import (
+	"context"
+	"time"
+
+	"emperror.dev/errors"
+	"github.com/aws/aws-sdk-go/service/elbv2"
+	"go.uber.org/cadence/activity"
+
+	"github.com/banzaicloud/pipeline/pkg/backoff"
+)
+
+const (
+	HealthCheckActivityName = "nlb-health-check"
+	backoffDelay            = 3 * time.Second
+	backoffMaxRetries       = 3
+)
+
+type HealthCheckActivity struct {
+	awsSessionFactory AWSSessionFactory
+	elbv2Factory      ELBV2APIFactory
+}
+
+// NewHealthCheckActivity returns a new HealthCheckActivity.
+func NewHealthCheckActivity(
+	awsSessionFactory AWSSessionFactory,
+	elbv2Factory ELBV2APIFactory,
+) *HealthCheckActivity {
+	return &HealthCheckActivity{
+		awsSessionFactory: awsSessionFactory,
+		elbv2Factory:      elbv2Factory,
+	}
+}
+
+type HealthCheckActivityInput struct {
+	OrganizationID uint
+	SecretID       string
+	Region         string
+	ClusterID      uint
+	ClusterName    string
+}
+
+// HealthCheckActivityOutput holds the output data
+type HealthCheckActivityOutput struct {
+	nlbStatus []*elbv2.TargetHealthDescription
+}
+
+// Register registers the activity in the worker.
+func (a *HealthCheckActivity) Register() {
+	activity.RegisterWithOptions(a.Execute, activity.RegisterOptions{Name: HealthCheckActivityName})
+}
+
+func (a *HealthCheckActivity) Execute(ctx context.Context, input HealthCheckActivityInput) (*HealthCheckActivityOutput, error) {
+	if a == nil {
+		return nil, errors.New("activity is nil")
+	}
+	session, err := a.awsSessionFactory.New(input.OrganizationID, input.SecretID, input.Region)
+	if err != nil {
+		return nil, errors.WrapIf(err, "failed to create AWS session")
+	}
+
+	elbv2Svc := a.elbv2Factory.New(session)
+	nlbName := "pke-" + input.ClusterName + "-nlb"
+
+	backoffConfig := backoff.ConstantBackoffConfig{
+		Delay:      backoffDelay,
+		MaxRetries: backoffMaxRetries,
+	}
+	backoffPolicy := backoff.NewConstantBackoffPolicy(backoffConfig)
+
+	var loadBalancerOutput *elbv2.DescribeLoadBalancersOutput
+	if err := backoff.Retry(func() error {
+		var errInt error
+		loadBalancerOutput, errInt = elbv2Svc.DescribeLoadBalancers(
+			&elbv2.DescribeLoadBalancersInput{
+				Names: []*string{&nlbName},
+			})
+		if errInt != nil {
+			return errors.WrapIf(errInt, "failed to describe load balancers")
+		}
+
+		return nil
+	}, backoffPolicy); err != nil {
```

Comment:
**Optional**: this call with the anonymous function definition would **MAY** be more readable preceding the if function instead of being inlined, because it does 3 things at the moment.

1. executes the backoff retry call.
2. defines the inlined anonymous function to describe the load balancers,
3. checks the error condition.

I would even consider extracting the anonymous function to a variable for readability purposes, but it is a secondary opinion.

Same at
https://github.com/banzaicloud/pipeline/pull/3229/files#diff-07dfcca59ca7ddb4d7bf740160fa2bdc2bc1668420e7549b77e0106c15a728e1R103
https://github.com/banzaicloud/pipeline/pull/3229/files#diff-07dfcca59ca7ddb4d7bf740160fa2bdc2bc1668420e7549b77e0106c15a728e1R119

Reply:
fixed: https://github.com/banzaicloud/pipeline/pull/3229/commits/93d075890463a1f969c8f4469a4a31a713cebbd3